rope.c

```diff
@@ -131,6 +131,85 @@ size_t rope_write_cstr(rope *r, uint8_t *dest) {
   return num_bytes + 1;
 }
 
+static inline uint8_t *skip_utf8(uint8_t *str, size_t chars, size_t bytelen);
+
+static inline size_t copy_utf8(uint8_t *dst, const uint8_t *src, size_t *chars,
+                               size_t max_size);
+
+size_t rope_write_substr(rope *rope, uint8_t *buf, size_t *bufsize,
+                         size_t from, size_t to) {
+  size_t len = rope_char_count(rope);
+  if (to > len) {
+    to = len;
+  }
+
+  rope_node *node = &rope->head;
+  size_t skipped = 0;
+
+  // Find the starting node.
+  while (skipped + node->nexts[0].skip_size < from) {
```

Comment:
This should use the same code as insert / delete. This has linear scan time, but the skip list structure allows log(n) time by using the other entries in the nexts pointer array.

Reply:
Oh wait, I see. Yeah this is log(n) but .. this could use `iter_at_char_pos` to find the location to read from instead. I guess thats more work, but less duplication. Hmm.